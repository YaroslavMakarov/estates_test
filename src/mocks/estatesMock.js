export const estatesMock = {
    estates: {
        estates: [
            {
                'id': 1,
                'title': 'Shevchenko estate',
                'price': 30000,
                'address': {
                    'street': 'Melnychna',
                    'number': 12
                },
                'preview_img': 'https://cp12.nevsepic.com.ua/62/1353962605-0435309-www.nevsepic.com.ua.jpg',
                'seller': 'Chack Norris',
                'description': 'The best from Chuck',
                'detailed_photos': [
                    'https://www.wtoc.com/resizer/ckPwhlUdYaFcLiBWMX0J033a0Ow=/1200x0/arc-anglerfish-arc2-prod-raycom.s3.amazonaws.com/public/JGJB35IW4JAMROXX3D3MMEF2SU.jpg',
                    'https://img3.goodfon.ru/wallpaper/nbig/c/f7/chuck-norris-chak-norris-legenda-aktior-vzgliad.jpg',
                    'https://cdn.flickeringmyth.com/wp-content/uploads/2020/03/chuck-norris.jpg'
                ]
            },
            {
                "id": 2,
                "title": "Yanukovich estate",
                "price": 3000000,
                "address": {
                    "street": "Prisident",
                    "number": 4
                },
                "preview_img": "https://biz.liga.net/images/general/2018/11/15/thumbnail-tw-20181115111659-8474.jpg?v=1542279867",
                "seller": "Chack Norris",
                "description": "The best from Chuck",
                "detailed_photos": [
                    "https://www.wtoc.com/resizer/ckPwhlUdYaFcLiBWMX0J033a0Ow=/1200x0/arc-anglerfish-arc2-prod-raycom.s3.amazonaws.com/public/JGJB35IW4JAMROXX3D3MMEF2SU.jpg",
                    "https://img3.goodfon.ru/wallpaper/nbig/c/f7/chuck-norris-chak-norris-legenda-aktior-vzgliad.jpg",
                    "https://cdn.flickeringmyth.com/wp-content/uploads/2020/03/chuck-norris.jpg"
                    ]
             },
        ],
    },
};