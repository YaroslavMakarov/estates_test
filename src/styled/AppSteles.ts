import styled from 'styled-components';

export const AppWrapper = styled.div`
    padding-top: 50px; 
`;