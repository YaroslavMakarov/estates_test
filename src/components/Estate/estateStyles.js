import styled from "styled-components";

export const EstateContainer = styled.div`
    box-sizing: border-box;
    padding: 15px;
`;
export const Img = styled.img.attrs(props => ({
    src: props.src
}))`
    display: block;
    width: 100%;
    height: 300px;
    object-fit: cover;
    margin-bottom: 10px;
`;
export const EstateDescription = styled.div`
    font-size: 16px;
    margin-bottom: 5px;
`;