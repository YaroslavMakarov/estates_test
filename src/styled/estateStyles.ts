import styled from 'styled-components';

export const EstateContainer = styled.div`
    box-sizing: border-box;
    padding: 15px;
`;
export const Img = styled.img.attrs(props => ({
    src: props.src
}))`
    display: block;
    width: 100%;
    height: ${props => props.id ? "400px" : "300px"};;
    object-fit: cover;
    margin-bottom: 10px;
`;
export const EstateDescription = styled.div`
    font-size: ${props => props.id ? "22px" : "16px"};
    margin-bottom: ${props => props.id ? "10px" : "5px"};
`;