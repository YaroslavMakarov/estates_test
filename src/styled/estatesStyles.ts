import styled from "styled-components";

export const EstatesWrapper = styled.div`
    display: grid;
    grid-template-columns: repeat(4, 250px);
    grid-gap: 15px;
    width: fit-content;
`; 