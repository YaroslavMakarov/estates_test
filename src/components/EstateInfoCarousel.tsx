import React from 'react';
import { Row, Col,
         Carousel, CarouselItem } from 'react-bootstrap';
import { Img } from '../styled/estateStyles';   

type Props = {
    photos: Array<string>;
};

const EstateInfoCarousel: React.FC<Props> = ({ photos }) => (
    <Row>
        <Col sm={{span: 6, offset: 3}}>
            <Carousel>
                {photos.map(photo => (
                    <CarouselItem
                        key={photo}
                    >
                        <Img src={photo} />
                    </CarouselItem>
                ))}
            </Carousel>
        </Col>
    </Row>
);

export default EstateInfoCarousel;
