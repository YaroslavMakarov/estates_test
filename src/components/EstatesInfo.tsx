import React from 'react';
import { useParams } from 'react-router-dom';
import { Row, Col } from 'react-bootstrap';
import 'bootstrap/dist/css/bootstrap.min.css';
import { useSelector } from 'react-redux';

import BackButton from '../components/BackButton';
import EstateInfoCarousel from '../components/EstateInfoCarousel';
import NoInfo from '../components/NoInfo';

import { Img, EstateDescription } from '../styled/estateStyles';
import { estatesSelector } from '../redux/estates';

type Params = {
    id: string;
};

const EstateInfo = () => {
const params: Params = useParams();
const estates = useSelector(estatesSelector);

const currentEstate = estates.filter(estate => estate.id === Number(params.id));

if (currentEstate.length === 0) {
    return (
        <NoInfo />
    );
}

    return (
        <>
            <BackButton />
            <Row>
                <Col sm={5}>
                    <Img 
                        src={currentEstate[0].preview_img}
                        id={params.id}
                    />
                </Col>
                <Col sm={{offset: 1, span: 5}}>
                    <EstateDescription id={params.id}>
                        {`Name: ${currentEstate[0].title}`}
                    </EstateDescription>
                    <EstateDescription id={params.id}>
                        {`Price: ${currentEstate[0].price}$`}
                    </EstateDescription>
                    <EstateDescription id={params.id}>
                        {`Address: ${currentEstate[0].address.street}, ${currentEstate[0].address.number}`}
                    </EstateDescription>
                    <EstateDescription id={params.id}>
                        {`Seller: ${currentEstate[0].seller}`}
                    </EstateDescription>
                    <EstateDescription id={params.id}>
                        {`Description: ${currentEstate[0].description}`}
                    </EstateDescription>
                </Col>
            </Row>
            <EstateInfoCarousel photos={currentEstate[0].detailed_photos} />
        </>
    )
};

export default EstateInfo;