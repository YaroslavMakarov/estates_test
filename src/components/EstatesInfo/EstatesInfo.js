import React from 'react';
import { useParams, useHistory } from 'react-router-dom';
import { Container, Row, Col, Button } from 'react-bootstrap';
import 'bootstrap/dist/css/bootstrap.min.css';
import { useSelector } from 'react-redux';

import EstateInfoCarousel from '../EstateInfoCarousel/EstateInfoCarousel';

import { Img, EstateDescription } from '../../styled/estateStyles';
import { estatesSelector } from '../../redux/estates';

const EstateInfo = () => {
const params = useParams();
const history = useHistory();
const estates = useSelector(estatesSelector);

const currentEstate = estates.find(estate => {
    if (estate.id === Number(params.id)) {
        return estate;
    }
})

    return (
        <>
            <Container
              fluide
              className="p-0"
            >
                <Button
                variant="dark"
                className="mb-3"
                onClick={() => history.goBack()}
                >
                    Back
                </Button>
            </Container>
            <Row>
                <Col sm={5}>
                    <Img 
                        src={currentEstate.preview_img}
                        id={params.id}
                    />
                </Col>
                <Col sm={{offset: 1, span: 5}}>
                    <EstateDescription id={params.id}>
                        {`Name: ${currentEstate.title}`}
                    </EstateDescription>
                    <EstateDescription id={params.id}>
                        {`Price: ${currentEstate.price}$`}
                    </EstateDescription>
                    <EstateDescription id={params.id}>
                        {`Address: ${currentEstate.address.street}, ${currentEstate.address.number}`}
                    </EstateDescription>
                    <EstateDescription id={params.id}>
                        {`Seller: ${currentEstate.seller}`}
                    </EstateDescription>
                    <EstateDescription id={params.id}>
                        {`Description: ${currentEstate.description}`}
                    </EstateDescription>
                </Col>
            </Row>
            <EstateInfoCarousel photos={currentEstate.detailed_photos} />
        </>
    )
};

export default EstateInfo;