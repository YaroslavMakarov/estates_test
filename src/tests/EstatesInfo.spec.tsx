import React from 'react';
import { mount, ReactWrapper } from 'enzyme';
import { Provider } from 'react-redux';
import { MemoryRouter, Route } from 'react-router-dom';
import * as ReactRedux from 'react-redux';
import configureStore from 'redux-mock-store';

import EstatesInfo from '../components/EstatesInfo';
import EstateInfoCarousel from '../components/EstateInfoCarousel';
import { Img, EstateDescription } from '../styled/estateStyles';
import { estatesMock } from '../mocks/estatesMock';

const mockStore = configureStore();

describe('EstateInfo component', () =>{
    let wrapper: ReactWrapper;
    let useSelector: jest.SpyInstance<unknown, [selector: (state: unknown) => unknown, equalityFn?: ((left: unknown, right: unknown) => boolean) | undefined]>;
    const mockedStore = mockStore(estatesMock);

    beforeEach(() => {
        useSelector = jest.spyOn(ReactRedux, 'useSelector')
            .mockImplementation(f => f(mockedStore.getState()));
        console.log(useSelector);
    });    
        
    describe('without info EstateInfo component', () => {
        beforeEach(() => {wrapper = mount(
            <Provider store={mockedStore}>
                <MemoryRouter initialEntries={['/info/3']}>
                    <Route path='/info/:id' component={EstatesInfo} />
                </MemoryRouter>
            </Provider>
        )});

        it('shoud get snapShot estates info without estates', () => {                
            expect(wrapper.debug()).toMatchSnapshot();
        });
    });
    
    describe('with info EstateInfo component', () => {
        let info: Estate;
        beforeEach(() => {
            info = estatesMock.estates.estates[0];
            wrapper = mount(
                <Provider store={mockedStore}>
                    <MemoryRouter initialEntries={['/info/1']}>
                        <Route path='/info/:id' component={EstatesInfo} />
                    </MemoryRouter>
                </Provider>
        )});

        it('shoud get snapShot estates info without estates', () => {                
            expect(wrapper.debug()).toMatchSnapshot();
        });

        it('chek Img attr', () => {
            const img = wrapper.find(Img).at(0);
            
            expect(img.prop('src')).toBe(`${info.preview_img}`);
        });

        it('check text EstateDescription', () => {
            wrapper.find(EstateDescription).forEach((node, index) => {
                const nodeNames = ['Name', 'Price', 'Address', 'Seller', 'Description'];
                switch(index) {
                    case 0: expect(node.text()).toBe(`${nodeNames[index]}: ${info.title}`);
                    break;
                    case 1: expect(node.text()).toBe(`${nodeNames[index]}: ${info.price}$`);
                    break;
                    case 2: expect(node.text()).toBe(`${nodeNames[index]}: ${info.address.street}, ${info.address.number}`);
                    break;
                    case 3: expect(node.text()).toBe(`${nodeNames[index]}: ${info.seller}`);
                    break;
                    case 4: expect(node.text()).toBe(`${nodeNames[index]}: ${info.description}`);
                    break;
                }                
            });
        });
    });
});