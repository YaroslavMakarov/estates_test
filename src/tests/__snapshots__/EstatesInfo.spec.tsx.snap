// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EstateInfo component with info EstateInfo component shoud get snapShot estates info without estates 1`] = `
"<Provider store={{...}}>
  <MemoryRouter initialEntries={{...}}>
    <Router history={{...}}>
      <Route path=\\"/info/:id\\" component={[Function: EstateInfo]}>
        <EstateInfo history={{...}} location={{...}} match={{...}} staticContext={[undefined]}>
          <BackButton>
            <Container fluid={true} className=\\"p-0\\">
              <div className=\\"p-0 container-fluid\\">
                <Button variant=\\"dark\\" className=\\"mb-3\\" onClick={[Function: onClick]} active={false} disabled={false}>
                  <button onClick={[Function: onClick]} disabled={false} type=\\"button\\" className=\\"mb-3 btn btn-dark\\">
                    Back
                  </button>
                </Button>
              </div>
            </Container>
          </BackButton>
          <Row noGutters={false}>
            <div className=\\"row\\">
              <Col sm={5}>
                <div className=\\"col-sm-5\\">
                  <styled.img src=\\"https://cp12.nevsepic.com.ua/62/1353962605-0435309-www.nevsepic.com.ua.jpg\\" id=\\"1\\">
                    <img src=\\"https://cp12.nevsepic.com.ua/62/1353962605-0435309-www.nevsepic.com.ua.jpg\\" id=\\"1\\" className=\\"sc-gsTCUz hUELus\\" />
                  </styled.img>
                </div>
              </Col>
              <Col sm={{...}}>
                <div className=\\"col-sm-5 offset-sm-1\\">
                  <styled.div id=\\"1\\">
                    <div id=\\"1\\" className=\\"sc-dlfnbm kGxffL\\">
                      Name: Shevchenko estate
                    </div>
                  </styled.div>
                  <styled.div id=\\"1\\">
                    <div id=\\"1\\" className=\\"sc-dlfnbm kGxffL\\">
                      Price: 30000$
                    </div>
                  </styled.div>
                  <styled.div id=\\"1\\">
                    <div id=\\"1\\" className=\\"sc-dlfnbm kGxffL\\">
                      Address: Melnychna, 12
                    </div>
                  </styled.div>
                  <styled.div id=\\"1\\">
                    <div id=\\"1\\" className=\\"sc-dlfnbm kGxffL\\">
                      Seller: Chack Norris
                    </div>
                  </styled.div>
                  <styled.div id=\\"1\\">
                    <div id=\\"1\\" className=\\"sc-dlfnbm kGxffL\\">
                      Description: The best from Chuck
                    </div>
                  </styled.div>
                </div>
              </Col>
            </div>
          </Row>
          <EstateInfoCarousel photos={{...}}>
            <Row noGutters={false}>
              <div className=\\"row\\">
                <Col sm={{...}}>
                  <div className=\\"col-sm-6 offset-sm-3\\">
                    <Carousel slide={true} fade={false} controls={true} indicators={true} defaultActiveIndex={0} interval={5000} keyboard={true} pause=\\"hover\\" wrap={true} touch={true} prevIcon={{...}} prevLabel=\\"Previous\\" nextIcon={{...}} nextLabel=\\"Next\\">
                      <div onKeyDown={[Function (anonymous)]} onMouseOver={[Function (anonymous)]} onMouseOut={[Function (anonymous)]} onTouchStart={[Function (anonymous)]} onTouchMove={[Function (anonymous)]} onTouchEnd={[Function (anonymous)]} className=\\"carousel slide\\">
                        <ol className=\\"carousel-indicators\\">
                          <li className=\\"active\\" onClick={[Function (anonymous)]} />
                          <li className={[undefined]} onClick={[Function (anonymous)]} />
                          <li className={[undefined]} onClick={[Function (anonymous)]} />
                        </ol>
                        <div className=\\"carousel-inner\\">
                          <Transition in={true} onEnter={[Function (anonymous)]} onEntered={[Function (anonymous)]} addEndListener={[Function: transitionEnd]} mountOnEnter={false} unmountOnExit={false} appear={false} enter={true} exit={true} onEntering={[Function: noop]} onExit={[Function: noop]} onExiting={[Function: noop]} onExited={[Function: noop]}>
                            <CarouselItem className=\\"active\\">
                              <div className=\\"active carousel-item\\">
                                <styled.img src=\\"https://www.wtoc.com/resizer/ckPwhlUdYaFcLiBWMX0J033a0Ow=/1200x0/arc-anglerfish-arc2-prod-raycom.s3.amazonaws.com/public/JGJB35IW4JAMROXX3D3MMEF2SU.jpg\\">
                                  <img src=\\"https://www.wtoc.com/resizer/ckPwhlUdYaFcLiBWMX0J033a0Ow=/1200x0/arc-anglerfish-arc2-prod-raycom.s3.amazonaws.com/public/JGJB35IW4JAMROXX3D3MMEF2SU.jpg\\" className=\\"sc-gsTCUz esHetz\\" />
                                </styled.img>
                              </div>
                            </CarouselItem>
                          </Transition>
                          <Transition in={false} onEnter={[Function: noop]} onEntered={[Function: noop]} addEndListener={[Function: transitionEnd]} mountOnEnter={false} unmountOnExit={false} appear={false} enter={true} exit={true} onEntering={[Function: noop]} onExit={[Function: noop]} onExiting={[Function: noop]} onExited={[Function: noop]}>
                            <CarouselItem className=\\"\\">
                              <div className=\\"carousel-item\\">
                                <styled.img src=\\"https://img3.goodfon.ru/wallpaper/nbig/c/f7/chuck-norris-chak-norris-legenda-aktior-vzgliad.jpg\\">
                                  <img src=\\"https://img3.goodfon.ru/wallpaper/nbig/c/f7/chuck-norris-chak-norris-legenda-aktior-vzgliad.jpg\\" className=\\"sc-gsTCUz esHetz\\" />
                                </styled.img>
                              </div>
                            </CarouselItem>
                          </Transition>
                          <Transition in={false} onEnter={[Function: noop]} onEntered={[Function: noop]} addEndListener={[Function: transitionEnd]} mountOnEnter={false} unmountOnExit={false} appear={false} enter={true} exit={true} onEntering={[Function: noop]} onExit={[Function: noop]} onExiting={[Function: noop]} onExited={[Function: noop]}>
                            <CarouselItem className=\\"\\">
                              <div className=\\"carousel-item\\">
                                <styled.img src=\\"https://cdn.flickeringmyth.com/wp-content/uploads/2020/03/chuck-norris.jpg\\">
                                  <img src=\\"https://cdn.flickeringmyth.com/wp-content/uploads/2020/03/chuck-norris.jpg\\" className=\\"sc-gsTCUz esHetz\\" />
                                </styled.img>
                              </div>
                            </CarouselItem>
                          </Transition>
                        </div>
                        <SafeAnchor className=\\"carousel-control-prev\\" onClick={[Function (anonymous)]}>
                          <a className=\\"carousel-control-prev\\" onClick={[Function: handleClick]} role=\\"button\\" href=\\"#\\" onKeyDown={[Function: handleKeyDown]}>
                            <span aria-hidden=\\"true\\" className=\\"carousel-control-prev-icon\\" />
                            <span className=\\"sr-only\\">
                              Previous
                            </span>
                          </a>
                        </SafeAnchor>
                        <SafeAnchor className=\\"carousel-control-next\\" onClick={[Function (anonymous)]}>
                          <a className=\\"carousel-control-next\\" onClick={[Function: handleClick]} role=\\"button\\" href=\\"#\\" onKeyDown={[Function: handleKeyDown]}>
                            <span aria-hidden=\\"true\\" className=\\"carousel-control-next-icon\\" />
                            <span className=\\"sr-only\\">
                              Next
                            </span>
                          </a>
                        </SafeAnchor>
                      </div>
                    </Carousel>
                  </div>
                </Col>
              </div>
            </Row>
          </EstateInfoCarousel>
        </EstateInfo>
      </Route>
    </Router>
  </MemoryRouter>
</Provider>"
`;

exports[`EstateInfo component without info EstateInfo component shoud get snapShot estates info without estates 1`] = `
"<Provider store={{...}}>
  <MemoryRouter initialEntries={{...}}>
    <Router history={{...}}>
      <Route path=\\"/info/:id\\" component={[Function: EstateInfo]}>
        <EstateInfo history={{...}} location={{...}} match={{...}} staticContext={[undefined]}>
          <NoInfo>
            <Container className=\\"d-flex align-items-center justify-content-center\\" fluid={false}>
              <div className=\\"d-flex align-items-center justify-content-center container\\">
                <styled.div>
                  <div className=\\"sc-hKgILt gmArwI\\">
                    Sorry, don&#39;t have any information about this estate.
                  </div>
                </styled.div>
              </div>
            </Container>
          </NoInfo>
        </EstateInfo>
      </Route>
    </Router>
  </MemoryRouter>
</Provider>"
`;
